src/bootstrap/src/core/sanity.rs

```diff
@@ -311,6 +311,17 @@ than building it.
             .entry(*target)
             .or_insert_with(|| Target::from_triple(&target.triple));
 
+        // compiler-rt c fallbacks for wasm cannot be built with gcc
+        if target.contains("wasm")
+            && (build.config.optimized_compiler_builtins(*target)
+                || build.config.rust_std_features.contains("compiler-builtins-c"))
+        {
+            let is_clang = build.cc_tool(*target).is_like_clang();
+            if !is_clang {
+                panic!("only clang supports building c code for wasm targets");
+            }
+        }
```

Comment:
Mind adding some more guidance to the panic message so anyone who hits this doesn't have to dig too much? Maybe something like

```rust
                panic!(
                    "Clang is required to build C code for Wasm targets, got `{}` instead\n\
                    this is because compiler-builtins is configured to build C source. Either \
                    ensure Clang is used, or adjust this configuration.",
                    tool.path().display()
                );
```